contrib/lib/src/compression/fairing.rs

```diff
@@ -0,0 +1,150 @@
+/// to 2 in order to achieve fast compression with a compression ratio similar
+/// to gzip. When appropriate, brotli's text and font compression modes are
+/// used.
+///
+/// In the gzip compression mode (using the
+/// [flate2](https://github.com/alexcrichton/flate2-rs) crate), quality is set
+/// The Compression type implements brotli and gzip compression for responses in
+/// to the default (9) in order to have good compression ratio.
+///
+/// This fairing does not compress responses that already have a
+/// `Content-Encoding` header.
+///
+/// This fairing ignores the responses with a `Content-Type` matching any of
+/// the following default types:
+///
+/// - application/gzip
+/// - application/brotli
+/// - application/zip
+/// - image/*
+/// - application/wasm
+/// - application/binary
+///
+/// The excluded types can be changed changing the `compress.exclude` Rocket
+/// configuration property.
+///
+/// # Usage
+///
+/// To use, add the `brotli_compression` feature, the `gzip_compression`
+/// feature, or the `compression` feature (to enable both algorithms) to the
+/// `rocket_contrib` dependencies section of your `Cargo.toml`:
+///
+/// ```toml,ignore
+/// [dependencies.rocket_contrib]
+/// version = "*"
+/// default-features = false
+/// features = ["compression"]
+/// ```
+///
+/// Then, ensure that the compression [fairing](/rocket/fairing/) is attached to
+/// your Rocket application:
+///
+/// ```rust
+/// extern crate rocket;
+/// extern crate rocket_contrib;
+///
+/// use rocket_contrib::compression::Compression;
+///
+/// fn main() {
+///     rocket::ignite()
+///         // ...
+///         .attach(Compression::fairing())
+///         // ...
+///     # ;
+/// }
+/// ```
+pub struct Compression(());
+
+impl Compression {
+    /// Returns a fairing that compresses outgoing requests.
+    ///
+    /// ## Example
+    /// To attach this fairing, simply call `attach` on the application's
+    /// `Rocket` instance with `Compression::fairing()`:
+    ///
+    /// ```rust
+    /// extern crate rocket;
+    /// extern crate rocket_contrib;
+    ///
+    /// use rocket_contrib::compression::Compression;
+    ///
+    /// fn main() {
+    ///     rocket::ignite()
+    ///         // ...
+    ///         .attach(Compression::fairing())
+    ///         // ...
+    ///     # ;
+    /// }
+    /// ```
+    pub fn fairing() -> Compression {
+        Compression { 0: () }
+    }
+}
+
+impl Fairing for Compression {
+    fn info(&self) -> Info {
+        Info {
+            name: "Response compression",
+            kind: Kind::Attach | Kind::Response,
+        }
+    }
+
+    fn on_attach(&self, rocket: Rocket) -> Result<Rocket, Rocket> {
+        let mut ctxt = Context::new();
+        match rocket.config().get_slice("compress.exclude") {
+            Ok(excls) => {
+                let mut error = false;
+                let mut exclusions_vec = Vec::with_capacity(excls.len());
```

Comment:
This can probably be written a bit more succinctly, along the lines of:

```rust
let exclusions: Result<Vec<String>, ConfigError> = excls.map(|e| match e {
    Value::String(s) => Ok(s.clone()),
    _ => Err(ConfigError::BadType(...)),
}).collect();

// Handle Ok, Err variants for exclusions...
```

It might be relatively easy to combine the error handling as well:
```rust
let exclusions = rocket.config().get_slice(...).and_then(|excls| ...);
let ctxt = match exclusions {
    Ok(e) => Context::with_exclusions(e),
    Err(ConfigError::Missing(_)) => { /* ignore missing */ },
    Err(e) => { e.pretty_print(); warn!(...) },
};
```

These should only be taken as rough ideas, I know they don't work directly as-is and it might be trickier than this.

Reply:
I have not been able to do this, mutability error in `excls.map` inside the `match` branch.